Add day 22 article (#98)
scalacenter/scala-advent-of-code: docs/puzzles/day22.md

```scala
extension (x1: Int)
  infix def by (x2: Int): Dimension = Dimension(x1, x2)
```
> `infix` is a keyword in Scala that allows us to define a method
> that can be called infix.

#### Does a Dimension Fit in Some Boundary? (`isSubset`)

We can test if some dimension fits within another
by checking that their minimum and maximum values conform
to each other. We can model this for `Dimension` with a
member method `isSubset`:

```scala
...
  def isSubset(d: Dimension): Boolean =
    min >= d.min && max <= d.max
```

#### Does one Dimension `intersect` with Another?

Lets also add another method to `Dimension` to get the
part that intersects with another dimension, if it exists:
```scala
...
  infix def insersect(d: Dimension): Option[Dimension] =
    Option.when(max >= d.min && min <= d.max) {
      (min max d.min) by (max min d.max)
    }
```
> You can think of intersecting dimensions as getting the part of
> two lines that overlap.

#### What `size` is a Dimension?

Lastly we add another method to `Dimension` to get its size:
```scala
...
  def size: Int = max - min + 1
```

#### Does one Cuboid `intersect` with Another?

> Now that we know how to get the intersection of a single
> dimension, we can extend this to the intersection of two

Reviewer: ```suggestion
Now that we know how to get the intersection of two
dimensions, we can extend this to the intersection of two
```

cuboids, by asserting that there is an intersection in
each dimension. Here we add a method to `Cuboid`:

```scala
...
  infix def intersect(curr: Cuboid): Option[Cuboid] =
    for
      xs <- this.xs insersect curr.xs
      ys <- this.ys insersect curr.ys
      zs <- this.zs insersect curr.zs
    yield
      Cuboid(xs, ys, zs)
```

#### What is the `volume` of a Cuboid?

With the `size` of each dimension determined, we can add a method
to `Cuboid` to determine its `volume`, it is computed as a
`BigInt` to avoid numeric overflow:
```scala

> ...

Reviewer: ```suggestion
```

  case s"x=${DimensionOf(xs)},y=${DimensionOf(ys)},z=${DimensionOf(zs)}" => Cuboid(xs, ys, zs)
```
```scala
val DimensionOf: Parser[Dimension] =
  case s"${NumOf(begin)}..${NumOf(end)}" => begin by end
```
```scala
val NumOf: Parser[Int] =
  case s if s.matches(raw"-?\d+") => s.toInt
```

#### Parsing a `Step`

Finally we can parse a single `Step`:

```scala
val StepOf: Parser[Step] =
  case s"${CommandOf(command)} ${CuboidOf(cuboid)}" => Step(command, cuboid)
```

To parse all lines, we can use `linesIterator`:
```scala
val steps = input.linesIterator.map(StepOf)
```
:::info
Note that the above call to `.map` will call
`StepOf.apply` for each line, this may throw a
`MatchError` if the line is formatted incorrectly.
:::

### Solution of Part 1

For part one, we only run the steps that are in the
initialisation sequence, i.e. running all the first
steps while they fit within the
area `x=-50..50,y=-50..50,z=-50..50`.

We check that a cuboid is in the initialisation
sequence with the following:

```scala
def isInit(cuboid: Cuboid): Boolean =
  Seq(cuboid.xs, cuboid.ys, cuboid.zs).forall(_.isSubset(-50 by 50))
```

The final code for part 1 is then to run the steps

> unly while they fit the initialisation sequence, and then

Reviewer: ```suggestion
only while they fit the initialisation sequence, and then
```
